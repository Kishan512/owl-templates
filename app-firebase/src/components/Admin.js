import { Component } from "../../lib/owl";

export class Admin extends Component {
  template = "Admin";
}
