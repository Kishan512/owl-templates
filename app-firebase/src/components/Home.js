import { Component } from "../../lib/owl";

export class Home extends Component {
  template = "Home";
}
