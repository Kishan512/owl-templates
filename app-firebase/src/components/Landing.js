import { Component } from "../../lib/owl";

export class Landing extends Component {
  template = "Landing";
}
