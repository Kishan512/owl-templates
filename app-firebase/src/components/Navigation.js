import { Component, router } from "../../lib/owl";
import { SignOutButton } from "./SignOut";

export class Navigation extends Component {
  template = "Navigation";
  components = { Link: router.Link, SignOutButton };
}
