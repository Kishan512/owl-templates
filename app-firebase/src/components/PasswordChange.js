import { Component } from "../../lib/owl";

const INITIAL_STATE = {
  passwordOne: "",
  passwordTwo: "",
  error: null
};

export class PasswordChange extends Component {
  template = "PasswordChange";

  constructor(parent, props) {
    super(parent, props);
    this.state = Object.assign({}, INITIAL_STATE);
  }

  async onSubmit(event) {
    event.preventDefault();
    try {
      const state = this.state;
      await this.env.firebase.updatePassword(state.passwordOne);
      this.state = Object.assign({}, INITIAL_STATE);
    } catch (error) {
      this.state.error = error;
    }
  }

  get isInvalid() {
    const state = this.state;
    return state.passwordOne !== state.passwordTwo || state.passwordTwo === "";
  }
}
