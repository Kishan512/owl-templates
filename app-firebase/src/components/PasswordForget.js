import { Component, router } from "../../lib/owl";

const INITIAL_STATE = {
  email: "",
  error: null
};

export class PasswordForget extends Component {
  template = "PasswordForget";
  state = Object.assign({}, INITIAL_STATE);

  async onSubmit(event) {
    event.preventDefault();
    try {
      const state = this.state;
      await this.env.firebase.resetPassword(state.email);
      this.state = Object.assign({}, INITIAL_STATE);
    } catch (error) {
      this.state.error = error;
    }
  }
}

export class PasswordForgetLink extends Component {
  template = "PasswordForgetLink";
  components = { Link: router.Link };
}
