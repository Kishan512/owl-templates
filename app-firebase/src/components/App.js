import { Component } from "../../lib/owl";
import { Navigation } from "./Navigation";

export class App extends Component {
  template = "App";
  components = { Navigation };
}
