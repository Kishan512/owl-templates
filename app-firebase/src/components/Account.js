import { PasswordChange } from "./PasswordChange";
import { Component } from "../../lib/owl";

export class Account extends Component {
  template = "Account";
  components = { PasswordChange };
}
