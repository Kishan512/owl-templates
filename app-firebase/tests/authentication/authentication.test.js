import { App } from "../../src/components/App";
import { makeTestEnv, makeTestFixture, nextTick, editInput } from "../helpers";

let fixture;
let env;

beforeEach(async () => {
  fixture = makeTestFixture();
  env = await makeTestEnv();
});

afterEach(() => {
  fixture.remove();
  env.cleanup();
});

describe("Various authentication flows", () => {
  test("can sign up", async () => {
    const app = new App(env);
    await app.mount(fixture);

    // we should be in the landing page
    expect(fixture.innerHTML).toMatchSnapshot();
    expect(document.location.pathname).toBe("/");

    // clicking on Sign In
    const signIn = fixture.querySelector('a[href="/signin"]');
    signIn.click();
    await nextTick();
    expect(document.location.pathname).toBe("/signin");
    expect(fixture.innerHTML).toMatchSnapshot();

    // there is now a link to Sign up
    const signUp = fixture.querySelector('a[href="/signup"]');
    signUp.click();
    await nextTick();
    expect(document.location.pathname).toBe("/signup");
    expect(fixture.innerHTML).toMatchSnapshot();

    // filling the form
    const signUpButton = fixture.querySelector('button[type="submit"');
    expect(signUpButton.disabled).toBe(true);
    await editInput(fixture.querySelector('input[name="username"]'), "aaron");
    await editInput(
      fixture.querySelector('input[name="email"]'),
      "myemail@gmail.com"
    );
    await editInput(
      fixture.querySelector('input[name="passwordOne"]'),
      "abcdef"
    );
    expect(signUpButton.disabled).toBe(true);
    await editInput(
      fixture.querySelector('input[name="passwordTwo"]'),
      "abcdef"
    );
    expect(signUpButton.disabled).toBe(false);

    // and finally click on Sign Up
    expect(env.session.authUser).toBe(null);
    signUpButton.click();
    await nextTick();

    expect(document.location.pathname).toBe("/home");
    expect(env.session.authUser).not.toBe(null);
    expect(fixture.innerHTML).toMatchSnapshot();
  });

  test("can sign in", async () => {
    const app = new App(env);
    await app.mount(fixture);

    // we should be in the landing page
    expect(document.location.pathname).toBe("/");

    // clicking on Sign In
    const signIn = fixture.querySelector('a[href="/signin"]');
    signIn.click();
    await nextTick();
    expect(document.location.pathname).toBe("/signin");
    expect(fixture.innerHTML).toMatchSnapshot();

    const signInButton =  fixture.querySelector('button[type="submit"');
    expect(signInButton.disabled).toBe(true);

    await editInput(fixture.querySelector('input[name="email"]'), "abc@email.com");
    expect(signInButton.disabled).toBe(true);
    await editInput(
      fixture.querySelector('input[name="password"]'),
      "123456"
    );
    expect(signInButton.disabled).toBe(false);
    expect(env.session.authUser).toBe(null);
    signInButton.click();
    await nextTick();

    expect(env.session.authUser).not.toBe(null);
    expect(document.location.pathname).toBe("/home");
    expect(fixture.innerHTML).toMatchSnapshot();
  });

  test("can sign out from home page", async () => {
    env.session.authUser = { some: "user" };
    await env.router.navigate({ to: "HOME" });
    const app = new App(env);
    await app.mount(fixture);

    // we should be in the landing page, signed in
    expect(fixture.innerHTML).toMatchSnapshot();
    expect(document.location.pathname).toBe("/home");
    expect(env.session.authUser).not.toBe(null);

    // clicking on Sign Out button
    const signout = fixture.querySelector("button");
    signout.click();
    await nextTick();

    expect(fixture.innerHTML).toMatchSnapshot();
    expect(document.location.pathname).toBe("/");
    expect(env.session.authUser).toBe(null);
  });

  test("can sign out from landing page", async () => {
    env.session.authUser = { some: "user" };
    const app = new App(env);
    await app.mount(fixture);

    // we should be in the landing page, signed in
    expect(fixture.innerHTML).toMatchSnapshot();
    expect(document.location.pathname).toBe("/");
    expect(env.session.authUser).not.toBe(null);

    // clicking on Sign Out button
    const signout = fixture.querySelector("button");
    signout.click();
    await nextTick();

    expect(fixture.innerHTML).toMatchSnapshot();
    expect(document.location.pathname).toBe("/");
    expect(env.session.authUser).toBe(null);
  });

  test("can load page", async () => {
    env.session.authUser = { some: "user" };
    await env.router.navigate({ to: "LANDING" });
    const app = new App(env);
    await app.mount(fixture);

    // we should be in the landing page, signed in
    expect(fixture.innerHTML).toMatchSnapshot();

    // we simulate here the answer from firebase
    env.firebase.userListenerCB(env.session.authUser);

    await nextTick();
    // we should still be in landing page
    expect(fixture.innerHTML).toMatchSnapshot();
  });

  test("is redirected to sign in page, then to correct page", async () => {
    window.history.pushState({}, "blabla", "/account");
    expect(document.location.pathname).toBe("/account");

    // we restart the router to take the url into account, no pun intended
    await env.router.start();

    // should have been redirected to /signin
    expect(document.location.pathname).toBe("/signin");
    const app = new App(env);
    await app.mount(fixture);

    // filling the form
    await editInput(fixture.querySelector('input[name="email"]'), "abc@email.com");
    await editInput(
      fixture.querySelector('input[name="password"]'),
      "123456"
    );
    fixture.querySelector('button[type="submit"').click();
    await nextTick();
    // should have been redirected to /account
    expect(document.location.pathname).toBe("/account");
  });
});
