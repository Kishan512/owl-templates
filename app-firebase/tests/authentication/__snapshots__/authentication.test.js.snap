// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Various authentication flows can load page 1`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"router-link-active\\">Landing</a></li><li><a href=\\"/home\\">Home</a></li><li><a href=\\"/account\\">Account</a></li><li><a href=\\"/admin\\">Admin</a></li><li><button> Sign Out </button></li></ul></div><hr><div> Landing... </div></div>"`;

exports[`Various authentication flows can load page 2`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"router-link-active\\">Landing</a></li><li><a href=\\"/home\\">Home</a></li><li><a href=\\"/account\\">Account</a></li><li><a href=\\"/admin\\">Admin</a></li><li><button> Sign Out </button></li></ul></div><hr><div> Landing... </div></div>"`;

exports[`Various authentication flows can sign in 1`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"\\">Landing</a></li><li><a href=\\"/signin\\" class=\\"router-link-active\\">Sign In</a></li></ul></div><hr><div><h1>Sign In</h1><form><input name=\\"email\\" type=\\"text\\" placeholder=\\"Email Address\\"><input name=\\"password\\" type=\\"password\\" placeholder=\\"Password\\"><button type=\\"submit\\" disabled=\\"\\"> Sign In </button></form><p><a href=\\"/pw-forget\\">Forgot password?</a></p><p> Don't have an account? <a href=\\"/signup\\">Sign Up</a></p></div></div>"`;

exports[`Various authentication flows can sign in 2`] = `"<div><div><ul><li><a href=\\"/\\">Landing</a></li><li><a href=\\"/home\\" class=\\"router-link-active\\">Home</a></li><li><a href=\\"/account\\">Account</a></li><li><a href=\\"/admin\\">Admin</a></li><li><button> Sign Out </button></li></ul></div><hr><div> Home... </div></div>"`;

exports[`Various authentication flows can sign out from home page 1`] = `"<div><div><ul><li><a href=\\"/\\">Landing</a></li><li><a href=\\"/home\\" class=\\"router-link-active\\">Home</a></li><li><a href=\\"/account\\">Account</a></li><li><a href=\\"/admin\\">Admin</a></li><li><button> Sign Out </button></li></ul></div><hr><div> Home... </div></div>"`;

exports[`Various authentication flows can sign out from home page 2`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"router-link-active\\">Landing</a></li><li><a href=\\"/signin\\">Sign In</a></li></ul></div><hr><div> Landing... </div></div>"`;

exports[`Various authentication flows can sign out from landing page 1`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"router-link-active\\">Landing</a></li><li><a href=\\"/home\\">Home</a></li><li><a href=\\"/account\\">Account</a></li><li><a href=\\"/admin\\">Admin</a></li><li><button> Sign Out </button></li></ul></div><hr><div> Landing... </div></div>"`;

exports[`Various authentication flows can sign out from landing page 2`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"router-link-active\\">Landing</a></li><li><a href=\\"/signin\\">Sign In</a></li></ul></div><hr><div> Landing... </div></div>"`;

exports[`Various authentication flows can sign up 1`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"router-link-active\\">Landing</a></li><li><a href=\\"/signin\\">Sign In</a></li></ul></div><hr><div> Landing... </div></div>"`;

exports[`Various authentication flows can sign up 2`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"\\">Landing</a></li><li><a href=\\"/signin\\" class=\\"router-link-active\\">Sign In</a></li></ul></div><hr><div><h1>Sign In</h1><form><input name=\\"email\\" type=\\"text\\" placeholder=\\"Email Address\\"><input name=\\"password\\" type=\\"password\\" placeholder=\\"Password\\"><button type=\\"submit\\" disabled=\\"\\"> Sign In </button></form><p><a href=\\"/pw-forget\\">Forgot password?</a></p><p> Don't have an account? <a href=\\"/signup\\">Sign Up</a></p></div></div>"`;

exports[`Various authentication flows can sign up 3`] = `"<div><div><ul><li><a href=\\"/\\" class=\\"\\">Landing</a></li><li><a href=\\"/signin\\" class=\\"\\">Sign In</a></li></ul></div><hr><div><h1>Sign Up</h1><form><input name=\\"username\\" type=\\"text\\" placeholder=\\"Full Name\\"><input name=\\"email\\" type=\\"text\\" placeholder=\\"Email\\"><input name=\\"passwordOne\\" type=\\"password\\" placeholder=\\"Password\\"><input name=\\"passwordTwo\\" type=\\"password\\" placeholder=\\"Password\\"><button type=\\"submit\\" disabled=\\"\\"> Sign Up </button></form></div></div>"`;

exports[`Various authentication flows can sign up 4`] = `"<div><div><ul><li><a href=\\"/\\">Landing</a></li><li><a href=\\"/home\\" class=\\"router-link-active\\">Home</a></li><li><a href=\\"/account\\">Account</a></li><li><a href=\\"/admin\\">Admin</a></li><li><button> Sign Out </button></li></ul></div><hr><div> Home... </div></div>"`;
